cardano-testnet/test/Test/Cli/QuerySlotNumber.hs

```diff
@@ -0,0 +1,157 @@
+      passedTime = fromIntegral passedSlots * fromNominalDiffTimeMicro sgSlotLength
+
+  poolNode1 <- H.headM poolNodes
+  poolSprocket1 <- H.noteShow $ nodeSprocket $ poolRuntime poolNode1
+  env <- H.evalIO getEnvironment
+  execConfig <- H.noteShow H.ExecConfig
+    { H.execConfigEnv = Last $ Just $
+      [ ("CARDANO_NODE_SOCKET_PATH", IO.sprocketArgumentName poolSprocket1)
+      ]
+      -- The environment must be passed onto child process on Windows in order to
+      -- successfully start that process.
+      <> env
+    , H.execConfigCwd = Last $ Just tempBaseAbsPath
+    }
+
+  -- {{{ Test cases
+  id do
+    H.note_ "Double check that start time offset is larger than the passed time we test"
+    H.diff startTimeOffsetSeconds (>) passedTime
+
+  id do
+    H.note_ "Try to retrieve slot 5s before genesis"
+    testTime <- H.note . formatTime $ (-5) `DT.addUTCTime` startTime
+    (result, _) <- H.runTestT $ execCli' execConfig
+      [ "query", "slot-number"
+      , "--testnet-magic", show @Int testnetMagic
+      , testTime
+      ]
+    H.assert $ isLeft result
+
+  id do
+    H.note_ "Retrieve slot number for the start time"
+    testTime <- H.note $ formatTime startTime
+    let expectedSlot = 0
+    slot <- readM @Int =<< execCli' execConfig
+      [ "query", "slot-number"
+      , "--testnet-magic", show @Int testnetMagic
+      , testTime
+      ]
+    isWithinDisk slot expectedSlot slotPrecision
+
+  id do
+    H.note_ "Retrieve slot number for some delay"
+    testTime <- H.note . formatTime $ passedTime `DT.addUTCTime` startTime
+    let expectedSlot = passedSlots
+    slot <- readM @Int =<< execCli' execConfig
+      [ "query", "slot-number"
+      , "--testnet-magic", show @Int testnetMagic
+      , testTime
+      ]
+    isWithinDisk slot expectedSlot slotPrecision
+
+  id do
+    H.note_ "Retrieve slot number for next epoch"
+    let epochDuration = slotLength * fromIntegral epochSize
+    testTime <- H.note . formatTime $ epochDuration `DT.addUTCTime` startTime
+    let expectedSlot = epochSize
+    slot <- readM @Int =<< execCli' execConfig
+      [ "query", "slot-number"
+      , "--testnet-magic", show @Int testnetMagic
+      , testTime
+      ]
+    isWithinDisk slot expectedSlot slotPrecision
+
+  id do
+    H.note_ "Try to retrieve slot far in the future"
+    -- 10 epochs ahead
+    let timeOffset = slotLength * fromIntegral epochSize * 10
+    testTime <- H.note . formatTime $ (startTimeOffsetSeconds + timeOffset) `DT.addUTCTime` startTime
+    (result, _) <- H.runTestT $ execCli' execConfig
+      [ "query", "slot-number"
+      , "--testnet-magic", show @Int testnetMagic
+      , testTime
+      ]
+    H.assert $ isLeft result
+
+  -- }}}
+
+formatTime :: DT.UTCTime -> String
+formatTime = DT.formatTime DT.defaultTimeLocale "%Y-%m-%dT%H:%M:%SZ"
+
+readM :: (Read a, Show a, H.MonadTest m, MonadCatch m, HasCallStack) => String -> m a
+readM = withFrozenCallStack . H.noteShowM . H.evalEither . readEither . trim
```

Comment:
Can you put this in `hedgehog-extras`?

Reply:
https://github.com/input-output-hk/hedgehog-extras/pull/32